packages/syft/src/syft/types/result.py

```diff
@@ -0,0 +1,98 @@
+# stdlib
+from collections.abc import Callable
+import functools
+from typing import Generic
+from typing import Literal
+from typing import NoReturn
+from typing import TypeAlias
+from typing import TypeVar
+
+T = TypeVar("T", covariant=True)
+E = TypeVar("E", covariant=True, bound=BaseException)
+BE = TypeVar("BE", bound=BaseException)
+
+
+class Ok(Generic[T]):
+    __slots__ = ("value",)
+    __match_args__ = ("ok_value",)
+
+    def __init__(self, value: T):
+        self.value = value
+
+    def __repr__(self) -> str:
+        return repr(self.value)
+
+    @property
+    def ok_value(self) -> T:
+        return self.value
+
+    def err(self) -> None:
+        return None
+
+    def ok(self) -> T:
+        return self.value
+
+    def is_err(self) -> Literal[False]:
+        return False
+
+    def is_ok(self) -> Literal[True]:
+        return True
+
+    def unwrap(self) -> T:
+        return self.value
+
+
+class Err(Generic[E]):
+    __slots__ = ("value",)
+    __match_args__ = ("error_value",)
+
+    def __init__(self, value: E):
+        self.value = value
+
+    def __repr__(self) -> str:
+        return repr(self.value)
+
+    @property
+    def error_value(self) -> E:
+        return self.value
+
+    def err(self) -> E:
+        return self.value
+
+    def ok(self) -> None:
+        return None
+
+    def is_err(self) -> Literal[True]:
+        return True
+
+    def is_ok(self) -> Literal[False]:
+        return False
+
+    def unwrap(self) -> NoReturn:
+        if isinstance(self.value, BaseException):
+            raise self.value
+        raise TypeError("Error is not a BaseException")
+
+
+Result: TypeAlias = Ok[T] | Err[E]
+
+
+def catch(
+    *exceptions: type[BE],
```

Comment:
A bit worried about always having to pass all exception types, this can get verbose very fast. 

Example: a function like `UserCodeService._call` calls 10+ other internal syft functions, if all return their own error types, we need a `@catch` with 10+ different exception types above the `_call` function

Reply:
It is also one of my concerns.

With `catch`, we don't necessarily have to carry all exceptions from one level to the other. However, if we're not carrying the exceptions over, why are we using result anyway? :)

One option is to carry "major" subclasses that groups many types of exceptions. For example, `@catch(StashException, UserCodeException)` instead of the other 10+ exceptions that will be thrown down the stack.

Reply:
An alternative to exception subclasses is using Enums to define custom exception groups (as Rust does)
This is more flexible than a class tree because it let's you predefine lists of specific exceptions which can be referenced by a single name while still allowing static type checking.
Something like

```
class StashErrors(Enum):
    VALIDATION_ERROR = ValidationError
    DATABASE_ERROR = DatabaseError
    NETWORK_ERROR = NetworkError
    UNKNOWN_ERROR = Exception
```